File: frontend/src/core/alerts/environment.ts
/* Copyright 2026 Marimo. All rights reserved. */

import type { NotificationMessageData, schemas } from "@/core/kernel/messages";

export type EnvironmentState = schemas["EnvironmentState"];
export type EnvironmentOperation = schemas["EnvironmentOperation"];
export type EnvironmentSource = EnvironmentOperation["source"];

export function emptyEnvironmentState(): EnvironmentState {
  return { restart_required: false, operations: [] };
}

/** Apply a live delta after the connection's authoritative snapshot. */
export function reduceEnvironmentState(
  state: EnvironmentState,
  update: NotificationMessageData<"environment-operation">,
): EnvironmentState {
  const previous = state.operations.find(
    (operation) => operation.operation_id === update.operation_id,
  );
  const logs = { ...previous?.logs };
  for (const [name, content] of Object.entries(update.logs)) {
    logs[name] =
      update.log_mode === "replace" ? content : (logs[name] ?? "") + content;
  }

  const operation: EnvironmentOperation = {
    operation_id: update.operation_id,
    action: update.action,
    status: update.status,
    packages: update.packages,
    logs,
    source: update.source,
  };
  const operations = state.operations.filter(
    (item) =>
      (previous && update.status.kind === "running") ||
      item.status.kind === "running",
  );
  const index = operations.findIndex(
    (item) => item.operation_id === update.operation_id,
  );
  if (index === -1) {
    operations.push(operation);
  } else {
    operations[index] = operation;
  }

  return {
    restart_required:
      state.restart_required ||
      update.status.kind === "restart-required" ||
      Object.values(update.packages).includes("restart-required"),
    operations,
  };
}

File: marimo/_session/state/environment.py
# Copyright 2026 Marimo. All rights reserved.
from __future__ import annotations

from marimo._messaging.notification import (
    EnvironmentOperation,
    EnvironmentOperationNotification,
    EnvironmentState,
    OperationRestartRequired,
    OperationRunning,
)


def reduce_environment_state(
    state: EnvironmentState,
    notification: EnvironmentOperationNotification,
) -> EnvironmentState:
    """Retain active attempts, the latest result, and outstanding restarts."""
    operations = {
        attempt.operation_id: attempt for attempt in state.operations
    }
    previous = operations.get(notification.operation_id)
    if previous is None or not isinstance(
        notification.status, OperationRunning
    ):
        operations = {
            operation_id: attempt
            for operation_id, attempt in operations.items()
            if isinstance(attempt.status, OperationRunning)
        }
    operations[notification.operation_id] = _reduce_operation(
        previous, notification
    )
    return EnvironmentState(
        # A later mutation cannot confirm that a different kernel was launched.
        restart_required=(
            state.restart_required
            or isinstance(notification.status, OperationRestartRequired)
            or "restart-required" in notification.packages.values()
        ),
        operations=list(operations.values()),
    )


def _reduce_operation(
    state: EnvironmentOperation | None,
    notification: EnvironmentOperationNotification,
) -> EnvironmentOperation:
    """Replace package progress and apply changes to named output streams."""
    packages = dict(notification.packages)
    logs = dict(state.logs) if state is not None else {}
    for name, content in notification.logs.items():
        if notification.log_mode == "replace":
            logs[name] = content
        else:
            logs[name] = logs.get(name, "") + content

    return EnvironmentOperation(
        packages=packages,
        logs=logs,
        source=notification.source,
        operation_id=notification.operation_id,
        action=notification.action,
        status=notification.status,
    )
